app/scripts/controllers/permissions/background-api.js

```diff
@@ -147,20 +185,56 @@ export function getPermissionBackgroundApiMethods({
     addPermittedAccounts: (origin, accounts) =>
       addMoreAccounts(origin, accounts),
 
-    removePermittedAccount: (origin, account) => {
+    removePermittedAccount: (origin, address) => {
       const caip25Caveat = getCaip25Caveat(origin);
       if (!caip25Caveat) {
         throw new Error(
-          `Cannot remove account "${account}": No permissions exist for origin "${origin}".`,
+          `Cannot remove account "${address}": No permissions exist for origin "${origin}".`,
         );
       }
 
-      const existingAccounts = getEthAccounts(caip25Caveat.value);
-
-      const remainingAccounts = existingAccounts.filter(
-        (existingAccount) => existingAccount !== account,
+      const permittedAccounts = new Set();
+      Object.values(caip25Caveat.value.requiredScopes).forEach(
+        ({ accounts }) => {
+          accounts.forEach((account) => {
+            permittedAccounts.add(account);
+          });
+        },
+      );
+      Object.values(caip25Caveat.value.optionalScopes).forEach(
+        ({ accounts }) => {
+          accounts.forEach((account) => {
+            permittedAccounts.add(account);
+          });
+        },
       );
 
+      const existingAccounts = Array.from(permittedAccounts);
+
+      const internalAccount = accountsController.getAccountByAddress(address);
+
+      const remainingAccounts = existingAccounts.filter((existingAccount) => {
+        const {
+          address: existingAddress,
+          chain: { namespace: existingNamespace, reference: existingReference },
+        } = parseCaipAccountId(existingAccount);
+
+        const matchesExistingAccount = internalAccount.scopes.some((scope) => {
+          const { namespace, reference } = parseCaipChainId(scope);
+
+          if (
+            namespace !== existingNamespace ||
+            !isEqualCaseInsensitive(internalAccount.address, existingAddress)
+          ) {
+            return false;
+          }
+
+          return reference === '0' || reference === existingReference;
```

Comment:
`reference === '0'` is scuffed... we have the worst of both worlds here having to handle both `eip155:0` and `wallet:eip155`

Reply:
alas yes. Any recommendations as to how to proceed for now?